### ISSUE TITLE
 a suggestion - clipping

### ISSUE THREAD

**Issue by mrT23** [kakaobrain/torchlars#3]
Hello

i have made extensive testing of torchlars package this week.
i couldn't get, under no circumstance, improvement over SGD.
in some cases i got lower results, in other i got NaN

i compared a bit your code to apex LARC implementation
https://github.com/NVIDIA/apex/blob/master/apex/parallel/LARC.py
it seems similar, and the one main difference is the clipping option, which
is true by default in LARC.
with LARC, i was able to outperform SGD results, and the runs are more stable, no Nans for all learning rate (although LARC optimizar is a little slower).

I suggest adding clipping as an option.
to your consideration

**Comment by cmpark0126** [comment]
Thank you for reporting what you experience!

Adding clipping option is not difficult, but I think there's another reason for that issue.

Actually, when I train the ResNet-50 model with ImageNet data using `torchlars`, there is no problem occurs. (Hyperparameters I used is mentioned at README)

So, I ask you to do below:

**First, could you let me know what parameters you used when you got strange results like NaN?**
(batch size, learning rate before warm-up, learning rate after warm-up, lr scheduling policy, etc.)

**Second, could you try to train your model with LARC with no clipping option, and same hyperparameters mentioned above?** If there is no problem when you try this option, I think troubleshooting is needed.

If you provide the above things, it will be really helpful to enhance `torchlars` package!

**Comment by mrT23** [comment]
my work now is still unpublished, so i cant share all the training details yet.
basically, with 60 epochs or training and batch size of ~40,000 i am able to reach:
78.7% with LARC
77.7% with SGD
76.5% with LARS

my top LARC run was with a learning rate of 18 and trust_coeff of 1e-2.
i repeated the run without grad clip, and gon NaN.

for LARS, i tried about ~10 runs, with trus coeff of 1e-3 and 1e-2 and various initial learning rate, from 2 to 40. the best result was with trus coeff of 1e-3 and learning rate of 24.
an example on NaN configuration: lr=4 | trust_coef=0.01

from my impression, grad clip is a game-changer. you gave me nice technical support earlier this week, and this post is just trying to extend the favor and give back algorithmic feedback.

all the best

**Comment by cmpark0126** [comment]
Thank you for sharing a detail of your experiments, it will be helpful for us.

Actually, I agree that the clipping option developed by NVIDIA is cool algorithm, but in terms of the meaning of the adaptive learning rate, that is a little bit different from the original LARS. If it really needed option, we really consider that add clipping option into `torchlars`, but before that, we need to check the reason why NaN is occurring and why LARS doesn't work properly. 

I really appreciate that you suggest clipping option, and we promise that consider that option positively. 

P.S. I have a plan to check this issue on hands next week, so please keep this issue open. I'll share some progress if I get new results. Thanks!
